packages/starlight/components/SidebarPersister.astro

```diff
@@ -0,0 +1,72 @@
+---
+/*
+  This component is designed to wrap the tree of `<SidebarSublist>` components in the sidebar.
+
+  It does the following:
+  - Wraps the tree in an `<sl-sidebar-state-persist>` custom element
+  - Before the tree renders, adds an inline script which loads state and defines
+  the behaviour for the `<sl-sidebar-restore>` custom element.
+  - After the tree renders, adds an inline script which restores the sidebar scroll state.
+
+  Notes:
+  - On smaller viewports, restoring state is skipped as the sidebar is collapsed inside a menu.
+  - The state is parsed from session storage and restored.
+  - This is a progressive enhancement, so any errors are swallowed silently.
+*/
+
+import type { Props } from '../props';
+import { getSidebarHash } from '../utils/navigation';
+
+const hash = getSidebarHash(Astro.props.sidebar);
+
+declare global {
+	interface Window {
+		/** Restored scroll position. Briefly stored on the `window` global to pass between inline scripts. */
+		_starlightScrollRestore?: number;
+	}
+}
+---
+
+<sl-sidebar-state-persist data-hash={hash}>
+	<script is:inline>
+		(() => {
+			try {
```

Comment:
Is the try/catch needed? I don't see anythink here that is likely to throw.

Also, you can do:

```js
{
  const target  = ...
}
```

instead of the function wrapper on line 32, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/block#block_scoping_rules_with_let_const_class_or_function_declaration_in_strict_mode

Reply:
The `try`/`catch` is mainly to defend against the state loading in case `sessionStorage` is unavailable or for some reason it got some bad value stored to it so the `JSON.parse()` throws.

Good point about block scoping! Hadn’t considered that before for these inline scripts.